projects/RabbitMQ.Client/client/impl/RabbitMQActivitySource.cs

```diff
@@ -48,47 +48,48 @@ public static class RabbitMQActivitySource
         {
-            new KeyValuePair<string, object>(MessagingSystem, "rabbitmq"),
-            new KeyValuePair<string, object>(ProtocolName, "amqp"),
-            new KeyValuePair<string, object>(ProtocolVersion, "0.9.1")
+            new KeyValuePair<string, object?>(MessagingSystem, "rabbitmq"),
+            new KeyValuePair<string, object?>(ProtocolName, "amqp"),
+            new KeyValuePair<string, object?>(ProtocolVersion, "0.9.1")
         };
 
-        internal static Activity Send(string routingKey, string exchange, int bodySize,
+        internal static Activity? Send(string routingKey, string exchange, int bodySize,
             ActivityContext linkedContext = default)
         {
-            if (s_publisherSource.HasListeners())
+            if (!s_publisherSource.HasListeners())
             {
-                Activity activity = linkedContext == default
-                    ? s_publisherSource.StartRabbitMQActivity(
-                        UseRoutingKeyAsOperationName ? $"{routingKey} publish" : "publish",
-                        ActivityKind.Producer)
-                    : s_publisherSource.StartLinkedRabbitMQActivity(
-                        UseRoutingKeyAsOperationName ? $"{routingKey} publish" : "publish",
-                        ActivityKind.Producer, linkedContext);
-                if (activity?.IsAllDataRequested == true)
-                {
-                    PopulateMessagingTags("publish", routingKey, exchange, 0, bodySize, activity);
-                }
+                return null;
+            }
 
-                return activity;
+            Activity? activity = linkedContext == default
+                ? s_publisherSource.StartRabbitMQActivity(
+                    UseRoutingKeyAsOperationName ? $"{routingKey} publish" : "publish",
+                    ActivityKind.Producer)
+                : s_publisherSource.StartLinkedRabbitMQActivity(
+                    UseRoutingKeyAsOperationName ? $"{routingKey} publish" : "publish",
+                    ActivityKind.Producer, linkedContext);
+            if (activity != null && activity.IsAllDataRequested)
+            {
+                PopulateMessagingTags("publish", routingKey, exchange, 0, bodySize, activity);
             }
 
-            return null;
+            return activity;
+
         }
 
-        internal static Activity ReceiveEmpty(string queue)
+        internal static Activity? ReceiveEmpty(string queue)
         {
             if (!s_subscriberSource.HasListeners())
             {
                 return null;
             }
 
-            Activity activity = s_subscriberSource.StartRabbitMQActivity(
+            Activity? activity = s_subscriberSource.StartRabbitMQActivity(
                 UseRoutingKeyAsOperationName ? $"{queue} receive" : "receive",
                 ActivityKind.Consumer);
-            if (activity.IsAllDataRequested)
+            if (activity != null && activity.IsAllDataRequested)
```

Comment:
I seems activity will never be null here due to implementation detail of StartRabbitMQActivity. Consider to put some assert instead, for example
```c#
Debug.Assert(activity != null, "StartRabbitMQActivity returns null only in case of no active listeners, and this case is already handled above.");
```

Reply:
StartRabbitMQActivity can be null though no?

Reply:
Technically yes, but in fact StartRabbitMQActivity returns result of `s_subscriberSource.CreateActivity(...)` which returns null only in case when `s_subscriberSource.HasListeners()` returns false. And this case was already handled separately. 
So, it seems the null check that you added does not fix any bug and it's redundant. We can keep it for extra safety of course, if you prefer.